### PR TITLE
feat: repeatable attributes support

### PR DESCRIPTION
**Pull request by PawelSuwinski**
Support for repeatable attributes, for example:

```yaml
Book:
    attributes:
      - ORM\UniqueConstraint: { name: "isbn", columns: ["isbn"] }
      - ORM\UniqueConstraint: { name: "title", columns: ["title"] }
   properties:
      isbn: ~
      title: { range: "https://schema.org/Text" }
```

gives:

```php
/**
 * A book.
 *
 * @see https://schema.org/Book
 */
#[ORM\Entity]
#[ApiResource(types: ['https://schema.org/Book'], routePrefix: '/library')]
#[ORM\UniqueConstraint(name: 'isbn', columns: ['isbn'])]
#[ORM\UniqueConstraint(name: 'title', columns: ['title'])]
#[MyAttribute]
class Book
{
```

### CODE REVIEW

File: src/Model/Attribute.php [hunk 1]
```diff
@@ -27,6 +27,14 @@ final class Attribute
 
     public bool $append = true;
 
+    /*
+     * Custom explicitly configured attributes is not mergeable with next one
+     * but treated as repeated if given more than once.
+     *
+     * @see ApiPlatform\SchemaGenerator\Model\AddAttributeTrait
+     */
```
Comment (alanpoulain) [inline src/Model/Attribute.php:35]:
```suggestion
    /**
     * Custom explicitly configured attributes is not mergeable with next one
     * but treated as repeated if given more than once.
     *
     * @see \ApiPlatform\SchemaGenerator\Model\AddAttributeTrait
     */
```

File: src/AttributeGenerator/ConfigurationAttributeGenerator.php [hunk 2]
```diff
@@ -26,16 +26,37 @@ final class ConfigurationAttributeGenerator extends AbstractAttributeGenerator
...
+        $typeAttributes = $this->config['types'][$class->name()]['attributes'] ?? [[]];
+        $vocabAttributes = [[]];
         if ($class instanceof SchemaClass) {
-            $vocabConfig = $this->config['vocabularies'][$class->resource()->getGraph()->getUri()] ?? null;
+            $vocabAttributes = $this->config['vocabularies'][$class->resource()->getGraph()->getUri()]['attributes'] ?? [[]];
         }
 
+        $getAttributesNames = fn (array $config) => array_unique(array_map(
+            fn (array $v) => array_keys($v)[0],
+            array_filter($config, fn (array $v) => !empty($v))
```
Comment (alanpoulain) [inline src/AttributeGenerator/ConfigurationAttributeGenerator.php:37]: Why filtering empty values?

File: src/AttributeGenerator/ConfigurationAttributeGenerator.php [hunk 3]
```diff
@@ -26,16 +26,37 @@ final class ConfigurationAttributeGenerator extends AbstractAttributeGenerator
...
+        ));
+        $typeAttributesNames = $getAttributesNames($typeAttributes);
+        $vocabAttributesNames = $getAttributesNames($vocabAttributes);
+
+        $attribute = fn (string $name, $args) => new Attribute($name, $args + [
```
Comment (alanpoulain) [inline src/AttributeGenerator/ConfigurationAttributeGenerator.php:42]:
```suggestion
        $getAttribute = fn (string $name, array $args) => new Attribute($name, $args + [
```

File: tests/Command/GenerateCommandTest.php [hunk 4]
```diff
@@ -115,6 +117,17 @@ class Book
             , $book);
         $this->assertStringContainsString(<<<'PHP'
     #[ORM\OrderBy(name: 'ASC')]
+PHP
+            , $book);
+        // Generated attribute could be merged with next one that is configured one.
```
Comment (alanpoulain) [inline tests/Command/GenerateCommandTest.php:122]:
```suggestion
        // Generated attribute could be merged with next one that is a configured one.
```

File: src/Model/AddAttributeTrait.php [hunk 5]
```diff
@@ -18,14 +18,19 @@ trait AddAttributeTrait
...
                     $this->attributes[] = $attribute;
                 }
             } else {
                 $this->attributes = array_map(
                     fn (Attribute $attr) => $attr->name() === $attribute->name()
-                        ? new Attribute($attr->name(), array_merge($attr->args(), $attribute->args()))
+                        ? new Attribute($attr->name(), array_merge(
+                            $attr->args(),
+                            $attribute->args(),
+                            ['mergeable' => $attribute->mergeable]
```
Comment (alanpoulain) [inline src/Model/AddAttributeTrait.php:32]: I'm not sure to understand why it is needed?

File: src/AttributeGenerator/ConfigurationAttributeGenerator.php (hunk 3 above)
Comment (PawelSuwinski) [reply src/AttributeGenerator/ConfigurationAttributeGenerator.php:42]: phpstan error with array typing:
``` Line   AttributeGenerator/ConfigurationAttributeGenerator.php                         
 ------ ------------------------------------------------------------------------------- 
  52     Parameter #2 $args of closure expects array,                                   
         array<array<string>|string>|bool|int|Nette\PhpGenerator\Literal|string given.  
58     Parameter #2 $args of closure expects array,                                   
         array<array<string>|string>|bool|int|Nette\PhpGenerator\Literal|string given.  
```

File: src/Model/AddAttributeTrait.php (hunk 5 above)
Comment (PawelSuwinski) [reply src/Model/AddAttributeTrait.php:32]: Without it every configured attribute won't be merged with previously generated by other generators, for example I assume that having such configuration:
```yaml
types:
  Book:
    properties:
     reviews:
        attributes:
          ORM\OneToMany: { cascade: [persist, remove] }
```
 we expect this:
```php
#[ORM\OneToMany(targetEntity: 'App\Entity\Review', mappedBy: 'book', cascade: ['persist', 'remove'])]
```
not this (first one generated before `ConfigurationAttributeGenerator`):
```php
#[ORM\OneToMany(targetEntity: 'App\Entity\Review', mappedBy: 'book'])]
#[ORM\OneToMany(cascade: ['persist', 'remove'])]
```

but on any further attribute with same name given in configuration we don't want merging,
if repeated by explicit we assume that user gave another form of repeatable attribute, so just for 
an example (OneToMany is not repeatable, so it is not properly used below):

```yaml
  reviews:
        attributes:
          - ORM\OneToMany: { cascade: [persist, remove] }
          - ORM\OneToMany: { targetEntity: 'App\Entity\Review', cascade: [all] }
```
 we expect this:

```php
#[ORM\OneToMany(targetEntity: 'App\Entity\Review', mappedBy: 'book', cascade: ['persist', 'remove'])]
#[ORM\OneToMany(targetEntity: 'App\Entity\Review', cascade: ['all'])]
```
not this:
```php
#[ORM\OneToMany(targetEntity: 'App\Entity\Review', mappedBy: 'book', cascade: ['all'])]
```

File: tests/Command/GenerateCommandTest.php (hunk 4 above)
Comment (PawelSuwinski) [reply tests/Command/GenerateCommandTest.php:122]: ok

File: src/AttributeGenerator/ConfigurationAttributeGenerator.php (hunk 2 above)
Comment (PawelSuwinski) [reply src/AttributeGenerator/ConfigurationAttributeGenerator.php:37]: Misspelled , should be `isset(array_keys([$v]))[0]`, but after all shorter will be just to test `$config === [[]]`.

File: src/Model/AddAttributeTrait.php (hunk 5 above)
Comment (PawelSuwinski) [reply src/Model/AddAttributeTrait.php:32]: In other words it transparently preserve all current attributes feature, adding new or merging with  already generated one with the same name (before `ConfigurationAttributeGenerator`) and possibility of declaring a new form of same attribute as a repetition.

**Comment by alanpoulain** [comment]
Thank you @PawelSuwinski.

**Pull request merged**